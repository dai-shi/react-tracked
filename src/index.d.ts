import * as React from 'react';

type EqlFn<V> = (a: V, b: V) => boolean;

// container

export const createContainer: <S, D>(useValue: () => [S, D]) => {
  customContext: React.Context<S>;
  Provider: React.ComponentType;
  useTrackedState: () => S;
  useTracked: () => [S, D];
  useDispatch: () => D;
  useSelector: <V>(selector: (state: S) => V, equalityFn?: EqlFn<V>) => V;
};

// default context

export type ProviderType<S = unknown, D = unknown> = React.ComponentType<{
  useValue: () => [S, D];
}>;

export const defaultContext: React.Context<unknown>;
export const Provider: ProviderType;
export const useTrackedState: <S>() => S;
export const useTracked: <S, D>() => [S, D];
export const useDispatch: <D>() => D;
export const useSelector: <S, V>(selector: (state: S) => V, equalityFn?: EqlFn<V>) => V;
