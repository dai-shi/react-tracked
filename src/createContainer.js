import { createProvider, createCustomContext } from './Provider';
import { createUseTrackedState, createUseTracked } from './useTrackedState';
import { createUseDispatch } from './useDispatch';
import { createUseSelector } from './useSelector';

export const createContainer = (useValue) => {
  const customContext = createCustomContext();
  return {
    customContext,
    Provider: createProvider(customContext, useValue),
    useTrackedState: createUseTrackedState(customContext),
    useTracked: createUseTracked(customContext),
    useDispatch: createUseDispatch(customContext),
    useSelector: createUseSelector(customContext),
  };
};
