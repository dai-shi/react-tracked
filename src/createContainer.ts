/* eslint react/destructuring-assignment: off */

import {
  FC,
  createContext as createContextOrig,
  createElement,
} from 'react';
import {
  createContext,
} from 'use-context-selector';

import {
  useTrackedState as useTrackedStateOrig,
  useTracked as useTrackedOrig,
} from './useTrackedState';
import { useUpdate as useUpdateOrig } from './useUpdate';
import { useSelector as useSelectorOrig } from './useSelector';

const warningObject = new Proxy({}, {
  get() { throw new Error('Please use <Provider>'); },
  apply() { throw new Error('Please use <Provider>'); },
});

export const createContainer = <State, Update extends (...args: any[]) => any, Props>(
  useValue: (props: Props) => readonly [State, Update],
) => {
  const StateContext = createContext(warningObject as State);
  const UpdateContext = createContextOrig(warningObject as Update);
  const Provider: FC<Props> = (props) => {
    const [state, update] = useValue(props);
    return createElement(UpdateContext.Provider, { value: update },
      createElement(StateContext.Provider, { value: state }, props.children));
  };
  const useTrackedState = (
    opts?: Parameters<typeof useTrackedStateOrig>[1],
  ) => useTrackedStateOrig(StateContext, opts);
  const useTracked = (
    opts?: Parameters<typeof useTrackedOrig>[2],
  ) => useTrackedOrig(StateContext, UpdateContext, opts);
  const useUpdate = () => useUpdateOrig(StateContext, UpdateContext);
  const useSelector = <Selected>(
    selector: (state: State) => Selected,
  ) => useSelectorOrig(StateContext, selector);
  return {
    Provider,
    useTrackedState,
    useTracked,
    useUpdate,
    useSelector,
  } as const;
};
