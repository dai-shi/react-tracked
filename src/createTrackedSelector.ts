import {
  useCallback,
  useEffect,
  useMemo,
  useReducer,
  useRef,
} from 'react';
import { createProxy, isChanged } from 'proxy-compare';

import { useAffectedDebugValue } from './utils';

export const createTrackedSelector = <State>(
  useSelector: <Selected>(selector: (state: State) => Selected) => Selected,
) => {
  const useTrackedSelector = () => {
    const [, forceUpdate] = useReducer((c) => c + 1, 0);
    const affected = new WeakMap();
    const lastAffected = useRef<typeof affected>();
    const prevState = useRef<State>();
    const lastState = useRef<State>();
    useEffect(() => {
      lastAffected.current = affected;
      if (prevState.current !== lastState.current
        && isChanged(
          prevState.current,
          lastState.current,
          affected,
          new WeakMap(),
        )) {
        prevState.current = lastState.current;
        forceUpdate();
      }
    });
    const selector = useCallback((nextState: State) => {
      lastState.current = nextState;
      if (prevState.current
        && prevState.current !== nextState
        && lastAffected.current
        && !isChanged(
          prevState.current,
          nextState,
          lastAffected.current,
          new WeakMap(),
        )
      ) {
        // not changed
        return prevState.current;
      }
      prevState.current = nextState;
      return nextState;
    }, []);
    const state = useSelector(selector);
    if (typeof process === 'object' && process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useAffectedDebugValue(state, affected);
    }
    const proxyCache = useMemo(() => new WeakMap(), []); // per-hook proxyCache
    return createProxy(state, affected, proxyCache);
  };
  return useTrackedSelector;
};
