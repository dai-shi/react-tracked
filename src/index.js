export { createContainer } from './createContainer';

export { Provider, defaultContext } from './Provider';
export { useTrackedState, useTracked } from './useTrackedState';
export { useDispatch } from './useDispatch';
export { useSelector } from './useSelector';
