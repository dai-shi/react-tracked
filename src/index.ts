export { createContainer } from './createContainer';
export { createTrackedSelector } from './createTrackedSelector';
export { memo } from './memo';
export { getUntracked as getUntrackedObject } from 'proxy-compare';
