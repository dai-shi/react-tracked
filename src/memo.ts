import { createElement, memo as reactMemo, forwardRef } from 'react';
import { trackMemo } from 'proxy-compare';

import type {
  PropsWithChildren,
  NamedExoticComponent,
  ComponentType,
  ComponentProps,
  MemoExoticComponent,
} from 'react';

export function memo<P extends Record<string, unknown>>(
  Component: ComponentType<P>,
  propsAreEqual?: (
    prevProps: Readonly<PropsWithChildren<P>>,
    nextProps: Readonly<PropsWithChildren<P>>,
  ) => boolean,
): NamedExoticComponent<P>;

export function memo<T extends ComponentType<any>>(
  Component: T,
  propsAreEqual?: (
    prevProps: Readonly<ComponentProps<T>>,
    nextProps: Readonly<ComponentProps<T>>,
  ) => boolean,
): MemoExoticComponent<T>;

export function memo(Component: any, propsAreEqual?: any) {
  const WrappedComponent = forwardRef((props: any, ref: any) => {
    Object.values(props).forEach(trackMemo);
    return createElement(Component, { ...props, ref });
  });
  return reactMemo(WrappedComponent, propsAreEqual);
}
