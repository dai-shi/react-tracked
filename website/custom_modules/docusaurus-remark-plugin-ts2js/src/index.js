/* eslint @typescript-eslint/no-var-requires: off */

const ts = require('typescript');
const prettier = require('prettier/standalone');
const parserTypescript = require('prettier/parser-typescript');
const prettierPluginEstree = require('prettier/plugins/estree');

const stripTypes = (origCode) => {
  const modifiedCode = origCode.replace(/\n\n/g, '\n//--EMPTYLINE--\n');
  const result = ts.transpileModule(modifiedCode, {
    compilerOptions: {
      target: ts.ScriptTarget.ES2018,
      module: ts.ModuleKind.ES2018,
      jsx: ts.JsxEmit.Preserve,
    },
  });
  const code = result.outputText;
  return code.replace(/\r/g, '').replace(/\n\/\/--EMPTYLINE--\n/g, '\n\n');
};

const format = (code) =>
  prettier.format(code, {
    parser: 'typescript',
    plugins: [parserTypescript, prettierPluginEstree],
    singleQuote: true,
    trailingComma: 'all',
  });

const uniqueId = 'UsedByRemarkPluginTs2Js'; // should this be configurable? maybe unnecessary.

const nodeForImport = {
  type: 'import',
  value: `import Tabs${uniqueId} from '@theme/Tabs';
import TabItem${uniqueId} from '@theme/TabItem';`,
};

const matchNode = (node) => node.type === 'code' && node.meta === 'ts2js';

// this returns a list to replace the node
const transformNode = (node) => {
  const jscode = format(stripTypes(node.value));
  const tscode = format(node.value);
  return [
    {
      type: 'jsx',
      value: `<Tabs${uniqueId}
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
    { label: 'TypeScript', value: 'ts', },
  ]}
>
<TabItem${uniqueId} value="js">`,
    },
    {
      type: node.type,
      lang: 'javascript',
      value: jscode,
    },
    {
      type: 'jsx',
      value: `</TabItem${uniqueId}>
<TabItem${uniqueId} value="ts">`,
    },
    {
      type: node.type,
      lang: 'typescript',
      value: tscode,
    },
    {
      type: 'jsx',
      value: `</TabItem${uniqueId}>
</Tabs${uniqueId}>`,
    },
  ];
};

module.exports = () => {
  let transformed = false;
  const transformer = (node) => {
    if (matchNode(node)) {
      transformed = true;
      return transformNode(node);
    }
    if (Array.isArray(node.children)) {
      let index = 0;
      while (index < node.children.length) {
        const result = transformer(node.children[index]);
        if (result) {
          node.children.splice(index, 1, ...result);
          index += result.length;
        } else {
          index += 1;
        }
      }
    }
    if (node.type === 'root' && transformed) {
      node.children.unshift(nodeForImport);
    }
    return null;
  };
  return transformer;
};
