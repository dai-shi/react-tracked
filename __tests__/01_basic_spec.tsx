import React, { useReducer, StrictMode } from 'react';

import { render, fireEvent, cleanup } from '@testing-library/react';

import { createContainer } from '../src/index';

describe('basic spec', () => {
  afterEach(cleanup);

  it('create a component', () => {
    const initialState = {
      count1: 0,
    };
    type Action = { type: 'increment' };
    const reducer = (state = initialState, action: Action) => {
      if (action.type === 'increment') {
        return { ...state, count1: state.count1 + 1 };
      }
      return state;
    };
    const useValue = () => useReducer(reducer, initialState);
    const { Provider, useTracked } = createContainer(useValue);
    const Counter = () => {
      const [state, dispatch] = useTracked();
      return (
        <div>
          <span>{state.count1}</span>
          <button type="button" onClick={() => dispatch({ type: 'increment' })}>+1</button>
        </div>
      );
    };
    const App = () => (
      <StrictMode>
        <Provider>
          <Counter />
          <Counter />
        </Provider>
      </StrictMode>
    );
    const { getAllByText, container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(getAllByText('+1')[0]);
    expect(container).toMatchSnapshot();
  });
});
