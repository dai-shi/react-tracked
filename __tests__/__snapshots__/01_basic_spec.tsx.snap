// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec create a component 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
</div>
`;

exports[`basic spec create a component 2`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
  <div>
    <span>
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
</div>
`;
